cuqi/problem/_problem.py

```diff
@@ -76,54 +135,60 @@ def get_components(cls, **kwargs):
 
         return problem.model, problem.data, problem_info
 
-    def __init__(self,likelihood,prior):
-        self.likelihood = likelihood
-        self.prior = prior
+    def __init__(self, *densities: Density, **data: np.ndarray):
+        self._target = JointDistribution(*densities)(**data)
+
+    def set_data(self, **kwargs) -> BayesianProblem:
+        """ Set the data of the problem. This conditions the underlying joint distribution on the data. """
+        if not isinstance(self._target, JointDistribution):
+            raise ValueError("Unable to set data for this problem. Maybe data is already set?")
+        self._target = self._target(**kwargs)
+        return self
 
     @property
     def data(self):
         """Extract the observed data from likelihood"""
         return self.likelihood.data
 
     @property
-    def likelihood(self):
+    def likelihood(self) -> Likelihood:
         """The likelihood function."""
-        return self._likelihood
-    
+        if not isinstance(self._target, Posterior):
+            raise ValueError(f"Unable to extract likelihood from this problem. Current target is: \n {self._target}")
```

Comment:
Is the term "target" defined? Maybe mention it in the doc string.

Reply:
Good point!